### PR TITLE
resources: convert user_config.js to TypeScript

### PR DESCRIPTION
**Pull request by quisquous**
This implements a few things from config that will probably need
to be moved there.

### CODE REVIEW

File: resources/user_config.ts [hunk 1]
```diff
@@ -242,12 +298,11 @@ class UserConfig {
...
             }
 
-            if (this.userFileCallbacks[overlayName])
-              this.userFileCallbacks[overlayName](jsFile, localFiles, options, basePath);
+            this.userFileCallbacks[overlayName]?.(jsFile, localFiles, options, basePath);
           } catch (e) {
             // Be very visible for users.
             console.log('*** ERROR IN USER FILE ***');
-            console.log(e.stack);
+            console.log(e);
```
Comment (quisquous) [inline resources/user_config.ts:305]: This is the only (intended) logic change.  TypeScript doesn't really support error types very well and `console.log(e)` covers this just as well.

File: types/data.d.ts [hunk 2]
```diff
@@ -2,10 +2,11 @@ import PartyTracker from '../resources/party.js';
 import { Lang } from './global';
 import { Job, Role } from './job';
 
-export interface Option {
+export interface BaseOptions {
```
Comment (quisquous) [inline types/data.d.ts:5]: I also changed this to `BaseOptions` with the idea that we'd have `RaidbossOptions extends BaseOptions` later.

File: resources/user_config.ts (hunk 1 above)
Comment (MaikoTan) [reply resources/user_config.ts:305]: What about `(e as Error).stack` ?

File: resources/user_config.ts [hunk 3]
```diff
@@ -25,21 +30,55 @@ const ZoneId = _ZoneId;
 import _ZoneInfo from './zone_info';
 const ZoneInfo = _ZoneInfo;
 
+// Convince TypeScript and eslint that these are used.  TypeScript doesn't have a great way
```
Comment (panicstevenson) [inline resources/user_config.ts:33]: This feels odd to me; none of `disable-next-line`, `disable-line`, or `disable` are preferable here?

File: resources/user_config.ts [hunk 4]
```diff
@@ -53,29 +92,41 @@ class UserConfig {
     //  * b/some_file.js
     //  * root_file1.js
     //  * root_file2.js
-    return keys.sort((a, b) => {
-      const maxLen = Math.max(splitKeyMap[a].length, splitKeyMap[b].length);
+    return keys.sort((keyA, keyB) => {
+      const listA = splitKeyMap[keyA];
+      const listB = splitKeyMap[keyB];
+      if (listA === undefined || listB === undefined)
+        return 0; // Convince TypeScript these exist.
```
Comment (panicstevenson) [inline resources/user_config.ts:99]: Which are we convincing here?

Comment (Trim21) [reply resources/user_config.ts:99]: ts think listA and listB maybe undefined

File: resources/user_config.ts (hunk 1 above)
Comment (quisquous) [reply resources/user_config.ts:305]: Yeah, I think that would work, but ultimately there's no interesting stack here (as compared to a runtime error).  The only stack is just "you're in the middle of loading a user file".

File: resources/user_config.ts (hunk 3 above)
Comment (quisquous) [reply resources/user_config.ts:33]: I can do `/* eslint-disable @typescript-eslint/no-unused-vars */` but then there is still a `ts(6133)` warning in VSCode (or an error when running normally).  I can use a `// @ts-ignore` to ignore the next line, but that ignores everything.  There's (deliberately??) no way to disable a single warning for a single line.

So, I thought "let me just use these" and then we at least get all the other errors still, but suppress the "unused" one.

File: resources/user_config.ts (hunk 4 above)
Comment (quisquous) [reply resources/user_config.ts:99]: > Which are we convincing here?

Sorry, I'm not sure what your question is.  Are you saying that these might actually be undefined?  Can you clarify what you're trying to get at?

Comment (panicstevenson) [reply resources/user_config.ts:99]: I don't know what "these" represents in the comment. It's a line comment, so I feel like "these" is the zero. But, that makes no sense, so I'm assuming it's listA and listB

Comment (quisquous) [reply resources/user_config.ts:99]: Moved the comment, thanks!

**Review by panicstevenson — APPROVED** [review]

**Pull request merged**